### PR TITLE
Use shell hacks to find valid PAT

### PR DESCRIPTION
**Pull request by supertopher**
if rate limit exceeded try another key

## What
we request too fast for github

## How
switch PATs when a rate limit is exceeded

## Recommended reading order
1. shell
2. YAML

image

### CODE REVIEW

**Comment by supertopher** [comment]
sample output:
```
Curl-ing rate_limit for PAT:  "REDACTED_0"
first 4 lines of Curl response:
{ "message": "Bad credentials", "documentation_url": "https://docs.github.com/rest" }


Running jq '.resources.core.remaining' on full_response
Github says this PAT has null left
./tools/bin/find_non_rate_limited_PAT: line 39: test: null: integer expression expected
Rate limit exceed for this PAT!
Curl-ing rate_limit for PAT:  "$REDACTED_1"
first 4 lines of Curl response:
{ "resources": { "core": { "limit": 5000, "used": 0, "remaining": 5000, "reset": 1648232311 }, "search": { "limit": 30, "used": 0, "remaining": 30, "reset": 1648228771 }, "graphql": { "limit": 5000, "used": 0, "remaining": 5000, "reset": 1648232311 }, "integration_manifest": { "limit": 5000, "used": 0, "remaining": 5000, "reset": 1648232311 }, "source_import": { "limit": 100, "used": 0, "remaining": 100, "reset": 1648228771 }, "code_scanning_upload": { "limit": 1000, "used": 0, "remaining": 1000, "reset": 1648232311 }, "actions_runner_registration": { "limit": 10000, "used": 0, "remaining": 10000, "reset": 1648232311 }, "scim": { "limit": 15000, "used": 0, "remaining": 15000, "reset": 1648232311 } }, "rate": { "limit": 5000, "used": 0, "remaining": 5000, "reset": 1648232311 } }


Running jq '.resources.core.remaining' on full_response
Github says this PAT has 5000 left
Curl-ing rate_limit for PAT:  "$REDACTED_1"
::set-output name=pat::$REDACTED_1
````

File: .github/actions/start-aws-runner/action.yml [hunk 1]
```diff
@@ -41,7 +41,7 @@ runs:
         aws-region: us-east-2
     - name: Start EC2 runner
       id: start-ec2-runner
-      uses: supertopher/ec2-github-runner@v15.0.5
+      uses: supertopher/ec2-github-runner@base64v1.0.5
```
Comment (supertopher) [inline .github/actions/start-aws-runner/action.yml:44]: I know I'm not exactly semantic here, but adding `base64` feels like a better clue for future readers

File: tools/bin/find_non_rate_limited_PAT [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    echo -e "$blue_text""Github says this PAT has ""$red_text""$rate_limit_remaining""$blue_text"" left""$default_text"
+
+    # If GH rate limit response > 100
+    if test $rate_limit_remaining -gt 100; then
+      # github actions will NOT pass a string that looks like a secret
+      base64_valid_pat=$(echo "$personal_access_token" | base64)
+      echo -e "$blue_text""Found a good PAT!!""$default_text"
+      # ::set-output name is a github action magic string for output
+      echo "::set-output name=pat::$base64_valid_pat"
```
Comment (supertopher) [inline tools/bin/find_non_rate_limited_PAT:44]: I tried sending this to `dev/null` but GitHub seems to need STDOUT to this steam to read the variable

**Review by supertopher — COMMENTED** [review]
@davinchia FYI
@timroes

**Review by pmossman — APPROVED** [review]
Sanity checked, looks good to me! Left a couple nits

File: tools/bin/find_non_rate_limited_PAT [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+# Run away from anything even a little scary
+set -o nounset # -u exit if a variable is not set
+set -o errexit # -f exit for any command failure
+# text color escape codes (please note \033 == \e but OSX doesn't respect the \e)
+blue_text='\033[94m'
+red_text='\033[31m'
+default_text='\033[39m'
+
+# set -x/xtrace' uses a Sony PS4 for more info
```
[the remark this answers is not in the document]
Comment (supertopher) [reply tools/bin/find_non_rate_limited_PAT:11]: I think they will be forever

[the remark this answers is not in the document]
Comment (tuliren) [reply tools/bin/find_non_rate_limited_PAT:11]: Just saw this today. LOL.

**Pull request merged**